core/common/src/main/java/alluxio/underfs/ObjectMultiRangeInputStream.java

```diff
@@ -0,0 +1,144 @@
+ */
+
+package alluxio.underfs;
+
+import alluxio.Configuration;
+import alluxio.PropertyKey;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A stream for reading data using range queries.
+ */
+@NotThreadSafe
+public abstract class ObjectMultiRangeInputStream extends InputStream {
+
+  /** Has the stream been closed. */
+  private boolean mClosed;
+  /** The backing input stream. */
+  protected InputStream mStream;
+  /** The current position of the stream. */
+  protected long mPos;
+  /** Position the current stream was open till (inclusive). */
+  protected long mEndPos;
+
+  @Override
+  public void close() throws IOException {
+    if (!mClosed) {
+      closeStream();
+    }
+    mClosed = true;
+  }
+
+  @Override
+  public int read() throws IOException {
+    if (mStream == null) {
+      openStream();
+    }
+    int value = mStream.read();
+    if (value != -1) { // valid data read
+      mPos++;
+      closeStreamIfBoundary();
+    }
+    return value;
+  }
+
+  @Override
+  public int read(byte[] b) throws IOException {
+    return read(b, 0, b.length);
+  }
+
+  @Override
+  public int read(byte[] b, int offset, int length) throws IOException {
+    if (length == 0) {
+      return 0;
+    }
+    if (mStream == null) {
+      openStream();
+    }
+    int read = mStream.read(b, offset, length);
+    if (read != -1) {
+      mPos += read;
+      closeStreamIfBoundary();
+    }
+    return read;
+  }
+
+  @Override
+  public long skip(long n) throws IOException {
+    if (n <= 0) {
+      return 0;
+    }
+    closeStream();
+    mPos += n;
+    openStream();
+    return n;
+  }
+
+  /**
+   * Close the current stream if the boundary for a range is crossed.
+   */
+  private void closeStreamIfBoundary() throws IOException {
+    if (mPos > mEndPos) {
+      closeStream();
+    }
+  }
+
+  /**
+   * Closes the current stream.
+   */
+  private void closeStream() throws IOException {
+    if (mStream == null) {
+      return;
+    }
+    mStream.close();
+    mStream = null;
+  }
+
+  /**
+   * Open a new stream reading a range.
+   *
+   * @param startPos start position in bytes (inclusive)
+   * @param endPos end position in bytes (inclusive)
+   * @return a new {@link InputStream}
+   * @throws IOException when a non-Alluxio error occurs
+   */
+  protected abstract InputStream createStream(long startPos, long endPos) throws IOException;
+
+  /**
+   * Block size for reading an object in chunks.
+   *
+   * @return block size in bytes
+   */
+  private long getBlockSize() {
+    return Configuration.getBytes(PropertyKey.USER_BLOCK_SIZE_BYTES_DEFAULT);
+  }
+
+  /**
+   * Opens a new stream at mPos if the wrapped stream mStream is null.
+   */
+  private void openStream() throws IOException {
+    if (mClosed) {
+      throw new IOException("Stream closed");
+    }
+    if (mStream != null) { // stream is already open
+      return;
+    }
```

Comment:
Does this remove the requirement of checking `mStream == null` before calling `openStream`?